include/swift/IDE/CodeCompletionResultType.h

```diff
@@ -107,54 +109,191 @@ enum class CodeCompletionResultTypeRelation : uint8_t {
   MAX_VALUE = Identical
 };
 
+class USRBasedType;
+
+/// Holds \c USRBasedTypes. Two \c USRBasedTypes with the same USR are
+/// represented by the same object if they live in the same arena.
+class USRBasedTypeArena {
+  friend class USRBasedType;
+
+  /// The allocator allocating the \c USRBasedTypes
+  llvm::BumpPtrAllocator Allocator;
+
+  /// Maps USRs to their \c USRBasedType instances.
+  llvm::StringMap<const USRBasedType *> CanonicalTypes;
+
+  /// Cache of the \c Void type because its frequently needed to compute type
+  /// relations. Set when \c getVoidType is called for the first time.
+  const USRBasedType *VoidType = nullptr;
+
+public:
+  /// Return the USR-based \c Void type. \p ASTCtx is needed to determine the
+  /// USR of the \c Void type.
+  const USRBasedType *getVoidType(const ASTContext &ASTCtx);
+
+  USRBasedTypeArena() {}
+};
+
+/// The equivalent of a \c ExpectedTypeContext to compute the type relation of
+/// \c USRBasedTypes.
+class USRBasedTypeContext {
+public:
+  class ContextualType {
+    /// The types that the result type must be identical/convertible to to
+    /// compute the type relation. Needs to be a vector because for the
+    /// conextual type `some MyProto & MyOtherProto`, the return type must be
+    /// convertible to both \c MyProto and \c MyOtherProto to be considered
+    /// convertible.
+    llvm::SmallVector<const USRBasedType *, 1> Types;
+
+    /// Whether a match against this type should be considered convertible
+    /// instead of identical. This is used to model optional conversions.
+    /// If we have
+    ///   let x: Int? = #^COMPLETE^#
+    /// we add both 'Int?' and 'Int' as contextual types to the
+    /// USRBasedTypeContext, but an identical match against 'Int' should only
+    /// be considered convertible.
+    bool IsConvertible;
+
+  public:
+    /// Compute the type relation of \p ResultType to this conextual type.
+    CodeCompletionResultTypeRelation
+    typeRelation(const USRBasedType *ResultType,
+                 const USRBasedType *VoidType) const;
+
+    ContextualType(ArrayRef<const USRBasedType *> Types, bool IsConvertible)
+        : Types(Types.begin(), Types.end()), IsConvertible(IsConvertible) {
+      assert(!Types.empty() && "USRBasedTypeContext::ContextualType should "
+                               "contain at least one type");
+    }
+  };
+
+private:
+  /// Cache of the \c Void type because its regularly used in type relation
+  /// computation.
+  const USRBasedType *VoidType;
```

Comment:
Makes sense. To do this, I needed to pre-compute the `Void` USR type when creating a `USRBasedTypeArena` at which point we don’t have an `ASTContext`. I thus hard-coded the USR of the `Void` type, which should be safe, I think because we don’t expect that to change.